src/coreclr/jit/valuenum.cpp

```diff
@@ -10617,13 +10617,16 @@ void Compiler::fgValueNumberTree(GenTree* tree)
                     CORINFO_CLASS_HANDLE handle    = gtGetClassHandle(addr, &isExact, &isNonNull);
                     if (isExact && (handle != NO_CLASS_HANDLE))
                     {
-                        // Make sure the "exact" handle is not shared
-                        if (info.compCompHnd->compareTypesForEquality(handle, handle) == TypeCompareState::Must)
+                        if (info.compCompHnd->canInlineTypeCheck(handle, CORINFO_INLINE_TYPECHECK_SOURCE_VTABLE) ==
+                            CORINFO_INLINE_TYPECHECK_PASS)
                         {
-                            JITDUMP("IND(obj) is actually a class handle for %s\n", eeGetClassName(handle));
-                            ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);
-                            tree->gtVNPair    = vnStore->VNPWithExc(ValueNumPair(handleVN, handleVN), addrXvnp);
-                            returnsTypeHandle = true;
+                            if (info.compCompHnd->compareTypesForEquality(handle, handle) == TypeCompareState::Must)
+                            {
+                                JITDUMP("IND(obj) is actually a class handle for %s\n", eeGetClassName(handle));
+                                ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);
```

Comment:
> `ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);`

This assumes the compile-time handle is the same as the runtime one, which is not true for AOT. The handle needs to be "embedded".

Reply:
@SingleAccretion thanks!! I keep forgetting about it 🙂 does it look good now? I see that there is a cache of handles in VN but in my case it's not guaranteed that the handle will be there + this path is taken rarely in AOT